sgkit/stats/pca.py

```diff
@@ -0,0 +1,341 @@
+    random_state: RandomStateType = 0,
+    variable: str = "call_alternate_allele_count",
+) -> BaseEstimator:
+    """ Create PCA estimator """
+    if ploidy is None:
+        if "ploidy" not in ds.dims:
+            raise ValueError(
+                "`ploidy` must be specified explicitly when not present in dataset dimensions"
+            )
+        ploidy = ds.dims["ploidy"]
+    if scaler is None:
+        scaler = "patterson"
+    if isinstance(scaler, str):
+        if scaler != "patterson":
+            raise ValueError(
+                f"Only 'patterson' scaler currently supported (not '{scaler}')"
+            )
+    if algorithm is None:
+        algorithm = "tsqr"
+    if algorithm not in ["tsqr", "randomized"]:
+        raise ValueError(
+            f"`algorithm` must be one of ['tsqr', 'randomized'] (not '{algorithm}')"
+        )
+
+    numblocks = da.asarray(_allele_counts(ds, variable, check_missing=False)).numblocks
+    if all(s > 1 for s in numblocks) and algorithm != "randomized":
+        raise ValueError(
+            "PCA can only be performed on arrays chunked in 2 dimensions if algorithm='randomized'. "
+            "Consider using this algorithm instead or rechunking the alternate allele counts array "
+            "(e.g. ds.call_alternate_allele_count.chunk((None, -1)))."
+        )
+
+    return Pipeline(
+        [
+            ("scaler", PattersonScaler(ploidy=ploidy)),
+            (
+                "pca",
+                TruncatedSVD(
+                    n_components=n_components,
+                    algorithm=algorithm,
+                    n_iter=n_iter,
+                    random_state=random_state,
+                    compute=False,
+                ),
+            ),
+        ]
+    )
+
+
+def pca_fit(
+    ds: Dataset,
+    est: BaseEstimator,
+    *,
+    variable: str = "call_alternate_allele_count",
+    check_missing: bool = True,
+) -> BaseEstimator:
+    """ Fit PCA estimator """
+    AC = _allele_counts(ds, variable, check_missing=check_missing)
+    return est.fit(da.asarray(AC).T)
+
+
+def pca_transform(
+    ds: Dataset,
+    est: BaseEstimator,
+    *,
+    variable: str = "call_alternate_allele_count",
+    check_missing: bool = True,
+    merge: bool = True,
+) -> Dataset:
+    """ Apply PCA estimator to new data """
+    AC = _allele_counts(ds, variable, check_missing=check_missing)
+    projection = est.transform(da.asarray(AC).T)
+    new_ds = Dataset({"sample_pca_projection": (("samples", "components"), projection)})
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+def _get(est: BaseEstimator, attr: str, fn: Any = lambda v: v) -> Optional[ArrayLike]:
+    try:
+        if hasattr(est, "named_steps"):
+            est = est["pca"]
+        return fn(getattr(est, attr))
+    except Exception:
+        return None
+
+
+def pca_stats(ds: Dataset, est: BaseEstimator, *, merge: bool = True) -> Dataset:
+    """ Extract attributes from PCA estimator """
```

Comment:
nit: when you have a single line docstring, you add a space prefix/suffix, why? Is this documented somewhere?

Reply:
https://github.com/pystatgen/sgkit/issues/325